lib/Conversion/TorchToTcp/Utils.cpp

```diff
@@ -76,3 +76,161 @@ Value torch_to_tcp::broadcastInLeadingDimsToMatchShape(
+  result = rewriter.create<tensor::ExpandShapeOp>(
+      result.getDefiningOp()->getLoc(), resultType, input, reassociationMap);
+  // Second: Broadcast Shape
+  // [1, 1, 1, 1] -> [N, C, H, W]
+  SmallVector<int64_t> axes;
+  SmallVector<Value> dimSizes;
+  for (int64_t axis = 0; axis < targetType.getRank(); ++axis) {
+    axes.push_back(axis);
+    dimSizes.push_back(rewriter.createOrFold<tensor::DimOp>(
+        result.getDefiningOp()->getLoc(), target, axis));
+  }
+  auto axesAttr = rewriter.getI64ArrayAttr(axes);
+  result = rewriter.create<tcp::BroadcastOp>(result.getDefiningOp()->getLoc(),
+                                             target.getType(), result, dimSizes,
+                                             axesAttr);
+
+  return result;
+}
+
+// Example: [C] -> [N, C, H, W]
+Value torch_to_tcp::broadcast1DToNDAndMatchShape(
+    ConversionPatternRewriter &rewriter, Value input, Value target,
+    int64_t axisInOutput) {
+
+  RankedTensorType inputType = input.getType().cast<RankedTensorType>();
+  RankedTensorType targetType = target.getType().cast<RankedTensorType>();
+
+  // This utility only accepts 1D inputs
+  assert(inputType.getRank() == 1);
+
+  Value result = input;
+
+  // First: Broadcast Rank
+  // [C] -> [1, C, 1, 1] if axisInOutput = 1
+  // reassociation map = [[0, 1, 2, 3]]
+  SmallVector<ReassociationExprs> reassociationMap(inputType.getRank());
+  for (int64_t axis = 0; axis < targetType.getRank(); ++axis)
+    reassociationMap[0].push_back(rewriter.getAffineDimExpr(axis));
+  SmallVector<int64_t> resultShape(targetType.getRank(), 1);
+  resultShape[axisInOutput] = inputType.getShape()[0];
+  auto resultType = targetType.cloneWith(makeArrayRef(resultShape),
+                                         targetType.getElementType());
+  result = rewriter.create<tensor::ExpandShapeOp>(
+      result.getDefiningOp()->getLoc(), resultType, input, reassociationMap);
+  // Second: Broadcast Shape
+  // [1, C, 1, 1] -> [N, C, H, W]
+  SmallVector<int64_t> axes;
+  SmallVector<Value> dimSizes;
+  for (int64_t axis = 0; axis < targetType.getRank(); ++axis) {
+    if (axis != axisInOutput) {
+      axes.push_back(axis);
+      dimSizes.push_back(rewriter.createOrFold<tensor::DimOp>(
+          result.getDefiningOp()->getLoc(), target, axis));
+    }
+  }
+  auto axesAttr = rewriter.getI64ArrayAttr(axes);
+  result = rewriter.create<tcp::BroadcastOp>(result.getDefiningOp()->getLoc(),
+                                             target.getType(), result, dimSizes,
+                                             axesAttr);
+
+  return result;
+}
+
+template <typename T>
+std::optional<Value>
+torch_to_tcp::getConstTensorUtil(PatternRewriter &rewriter, Operation *op,
+                                 ArrayRef<T> vec, ArrayRef<int64_t> shape,
+                                 RankedTensorType type) {
+  uint64_t numTotalElements = 1;
+  for (int64_t a : shape) {
+    numTotalElements *= a;
```

Comment:
Should there be a check here that the `shape` has only static sizes?

Reply:
Good point. Done, PTAL.
```
  for (int64_t a : shape) {
    if (a == mlir::torch::Torch::kUnknownSize) {
        op->emitOpError("getConstTensor(): only static sizes supported");
        return std::nullopt;
    }
    numTotalElements *= a;
  }
```

Reply:
I think assuming that `shape` was created from `Torch` shapes is fragile. At the level of the `TorchToTcp` pass, there will be torch types and builtin tensor types in use. I would instead simply check that `a` is positive.

Reply:
Replaced with an assert, since this utility creates `tcp.const` and should never be called with dynamic shapes. It is the caller's responsibility to ensure this.
```
assert(a >= 0 && "getConstTensor(): Only static shapes supported");
```